packages/nimble-components/src/wafer-map/specs/features/rendering.md

```diff
@@ -0,0 +1,168 @@
+
+The POC is found in this branch [Worker Rendering POC](https://github.com/ni/nimble/compare/main...users/munteannatan/scale_and_test_updates).
+
+### Data Structure and Interface
+
+This is the proposed representation of the wafer information in memory
+
+```TS
+class WaferData {
+    // the x coordinates of each column of dies
+    dieColIndexArray: Int32Array;
+    // the lengths of each row of dies
+    rowLengthsArray: Int32Array;
+    // the y coordinates of each die as a matrix row by row
+    dieRowIndexLayer: Int32Array;
+    // the value of each die as a matrix row by row
+    dieValuesLayer: Int32Array;
+    // the highlight state of each die as a matrix row by row
+    dieHighlightsLayer: Int8Array;
+    // extra values layers
+    extraLayers : {
+        name: string;
+        type: string;
+        buffer: TypedArray;
+    }[]
+}
+```
+
+It will be used to store the information and it can be transferred to a worker thread with minimal effort by copying the arrayBuffers and reconstructing the object.
+
+Other benefits include the low access time to the values and the possibility to extend the layers of values.
+
+The previous inputs can be adapted to this new structure to maintain backwards compatibility.
+
+### Rendering
+
+The rendering will also be changed to take place inside a worker thread. The worker will be responsible for parsing each layer of the data set and it will render the information contained with the specific color codes and opacity.
+
+We will be using the [threads.js](https://threads.js.org/) library to spin up the worker and communicate with it.
+
+The worker will function as a state automaton which will need some data after initialization, and it will render everything using an [OffscreenCanvas](https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas).
+
+The main thread will communicate with the worker by signaling any data changes or user events which will trigger full or partial renders.
+
+The strategy for improvement will be gradual. We will try to reuse the existing code as much as possible, but if the performance will not be met, we will improve the existing parsing methods and increase the number of workers.
+
+#### Size Limitations
+
+Because we are reaching out to support very large amounts of data, an issue that is popping out is the limits of the component's canvas size to the amount of displayed data. For example we could find ourselves in the situation to render more dies than there are pixels on the canvas, at least in the zoomed out state.
+
+The solution to this issue is creating a new validity warning or event which will signal this state to the user when the count of individual dies will overtake the size of the canvas. A heuristic which will calculate this based on the zoom level and the component height and width will be implemented and used during data validation.
```

Comment:
I'm wary of describing what you outline here as "the solution". It suggests that this is how we intend to handle this scenario if not ideally, than indefinitely, which I'm not convinced is the best answer. We discussed some UX and technical possibilities for improving this experience, from decimation of the data (server-side and possibly even client-side in workers), to providing a more instantaneous response to zoom gestures using bitmap scaling. I think these ideas should be captured here, and either have reasons for why we feel they will never be pursued or highlighted as future considerations.

Has the validity warning you describe been designed? I'd be curious to see what was envisioned for this.